data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeEvaluatorListener.java

```diff
@@ -0,0 +1,387 @@
+                    + getPartialStatementFromContext(ctx), e);
+        }
+    }
+
+    @Override
+    public void enterUnaryOperator(DataPrepperExpressionParser.UnaryOperatorContext ctx) {
+
+    }
+
+    @Override
+    public void exitUnaryOperator(DataPrepperExpressionParser.UnaryOperatorContext ctx) {
+
+    }
+
+    @Override
+    public void enterPrimary(DataPrepperExpressionParser.PrimaryContext ctx) {
+
+    }
+
+    @Override
+    public void exitPrimary(DataPrepperExpressionParser.PrimaryContext ctx) {
+
+    }
+
+    @Override
+    public void enterJsonPointer(DataPrepperExpressionParser.JsonPointerContext ctx) {
+
+    }
+
+    @Override
+    public void exitJsonPointer(DataPrepperExpressionParser.JsonPointerContext ctx) {
+
+    }
+
+    @Override
+    public void enterVariableIdentifier(DataPrepperExpressionParser.VariableIdentifierContext ctx) {
+
+    }
+
+    @Override
+    public void exitVariableIdentifier(DataPrepperExpressionParser.VariableIdentifierContext ctx) {
+
+    }
+
+    @Override
+    public void enterVariableName(DataPrepperExpressionParser.VariableNameContext ctx) {
+
+    }
+
+    @Override
+    public void exitVariableName(DataPrepperExpressionParser.VariableNameContext ctx) {
+
+    }
+
+    @Override
+    public void enterLiteral(DataPrepperExpressionParser.LiteralContext ctx) {
+
+    }
+
+    @Override
+    public void exitLiteral(DataPrepperExpressionParser.LiteralContext ctx) {
+
+    }
+
+    @Override
+    public void visitTerminal(TerminalNode node) {
+        final int nodeType = node.getSymbol().getType();
+        if (nodeType == DataPrepperExpressionParser.EOF) {
+            return;
+        }
+        if (strategy.containsKey(nodeType) || nodeType == DataPrepperExpressionParser.LPAREN) {
+            operatorSymbolStack.push(nodeType);
+        } else if (nodeType == DataPrepperExpressionParser.RPAREN) {
+            // pop LPAREN at operatorSymbolStack top
+            operatorSymbolStack.pop();
+        } else {
+            try {
+                final Object arg = coercionService.coerceTerminalNode(node, context);
+                operandStack.push(arg);
+            } catch (final CoercionException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+    }
+
+    @Override
+    public void visitErrorNode(ErrorNode node) {
+
+    }
+
+    @Override
+    public void enterEveryRule(ParserRuleContext ctx) {
+
+    }
+
+    @Override
+    public void exitEveryRule(ParserRuleContext ctx) {
+
+    }
+
+    private void performSingleOperation(final int numOfArgs) {
+        final Integer operatorType = operatorSymbolStack.pop();
+        final Operator<?> operator = strategy.get(operatorType);
+        final Object[] args = new Object[numOfArgs];
+        for (int i = numOfArgs - 1; i >= 0; i--) {
+            args[i] = operandStack.pop();
+        }
+        final Object result = operator.evaluate(args);
```

Comment:
Now that I see how this is used, I think `Operator::evaluate` could be a `List`. I'd generally only consider varargs when passing in arguments like:

```
operator.evaluate(lefthand, righthand);
```

and

```
operator.evaluate(unary);
```